File: sophron_swarm/checkpointer.py
"""
SQLite-backed append-only transaction log checkpointer (spec §2).

Each state transition is persisted as an immutable row, enabling granular
rollback to any prior SwarmState snapshot for thread recovery or debugging.

The log is strictly append-only: rows are never modified or deleted, providing
a complete immutable audit trail of every agent-driven state transition.
"""
from __future__ import annotations

import sqlite3
import threading
from datetime import datetime, timezone
from typing import Optional

from sophron_swarm.state import SwarmState


_DDL = """
CREATE TABLE IF NOT EXISTS checkpoints (
    seq        INTEGER PRIMARY KEY AUTOINCREMENT,
    thread_id  TEXT    NOT NULL,
    timestamp  TEXT    NOT NULL,
    bitmask    INTEGER NOT NULL,
    state_json TEXT    NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_thread_seq ON checkpoints (thread_id, seq);
"""


class Checkpointer:
    """
    Lightweight, thread-safe SQLite checkpointer.

    Uses one connection per thread via threading.local() to avoid contention.
    WAL journal mode is enabled for concurrent read performance.

    Parameters
    ----------
    db_path:
        SQLite database file path.  Use ":memory:" (default) for an in-process
        store, or provide an absolute path for persistence across restarts.
    """

    def __init__(self, db_path: str = ":memory:") -> None:
        self._db_path = db_path
        self._local = threading.local()
        self._init_error: str | None = None
        try:
            self._init_schema()
        except sqlite3.Error as exc:
            # Don't crash the entire application if the DB is unavailable;
            # save/load will gracefully degrade to no-ops.
            import logging
            logging.getLogger(__name__).warning(
                "Checkpointer init failed (%s) – running without persistence.", exc
            )
            self._init_error = str(exc)

    # ── Connection pool (one connection per thread) ───────────────────────────

    @property
    def _conn(self) -> sqlite3.Connection:
        if not getattr(self._local, "conn", None):
            conn = sqlite3.connect(self._db_path, check_same_thread=False)
            conn.execute("PRAGMA journal_mode=WAL;")
            conn.execute("PRAGMA synchronous=NORMAL;")
            self._local.conn = conn
        return self._local.conn

    def _init_schema(self) -> None:
        self._conn.executescript(_DDL)
        self._conn.commit()

    # ── Public API ────────────────────────────────────────────────────────────

    def save(self, thread_id: str, state: SwarmState) -> Optional[int]:
        """
        Append a new checkpoint row.

        Returns the auto-assigned sequence number, or None if the save failed
        (errors are logged but never raised, so checkpoint failures cannot
        crash the graph loop).
        """
        if self._init_error:
            return None
        try:
            cur = self._conn.execute(
                "INSERT INTO checkpoints(thread_id, timestamp, bitmask, state_json) "
                "VALUES (?, ?, ?, ?)",
                (
                    thread_id,
                    datetime.now(timezone.utc).isoformat(),
                    state.bitmask,
                    state.model_dump_json(),
                ),
            )
            self._conn.commit()
            return cur.lastrowid  # type: ignore[return-value]
        except sqlite3.Error as exc:
            import logging
            logging.getLogger(__name__).warning(
                "Checkpoint save failed: %s", exc
            )
            return None

    def load_latest(self, thread_id: str) -> Optional[SwarmState]:
        """Return the most recent SwarmState for this thread, or None."""
        if self._init_error:
            return None
        try:
            row = self._conn.execute(
                "SELECT state_json FROM checkpoints "
                "WHERE thread_id = ? ORDER BY seq DESC LIMIT 1",
                (thread_id,),
            ).fetchone()
            return SwarmState.model_validate_json(row[0]) if row else None
        except (sqlite3.Error, Exception) as exc:
            import logging
            logging.getLogger(__name__).warning(
                "load_latest failed: %s", exc
            )
            return None

    def load_at(self, thread_id: str, seq: int) -> Optional[SwarmState]:
        """Return the SwarmState at a specific sequence number (rollback)."""
        if self._init_error:
            return None
        try:
            row = self._conn.execute(
                "SELECT state_json FROM checkpoints WHERE thread_id = ? AND seq = ?",
                (thread_id, seq),
            ).fetchone()
            return SwarmState.model_validate_json(row[0]) if row else None
        except (sqlite3.Error, Exception) as exc:
            import logging
            logging.getLogger(__name__).warning(
                "load_at failed: %s", exc
            )
            return None

    def list_checkpoints(self, thread_id: str) -> list[dict]:
        """Return metadata for all checkpoints in a thread (no state JSON)."""
        rows = self._conn.execute(
            "SELECT seq, timestamp, bitmask FROM checkpoints "
            "WHERE thread_id = ? ORDER BY seq",
            (thread_id,),
        ).fetchall()
        return [
            {"seq": r[0], "timestamp": r[1], "bitmask": f"0x{r[2]:04X}"}
            for r in rows
        ]

File: src/state/checkpointer.ts
/**
 * Checkpointer — immutable, append-only SQLite state log.
 *
 * Ported from V2's sophron_swarm/checkpointer.py. One row per state transition.
 * WAL mode for concurrent read during writes. All failures degrade to no-ops
 * (logged, never crash the loop) — matches V2's `_safe_checkpoint`.
 *
 * See docs/PHASE_0_DESIGN.md §8.
 */
import Database from "better-sqlite3";
import { mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { log } from "../util/log.js";
import type { AgentRunState } from "../types.js";

export class Checkpointer {
  private db: Database.Database;

  constructor(dbPath = ".sophron/checkpoint.db") {
    try {
      mkdirSync(dirname(dbPath), { recursive: true });
    } catch {
      // dirname may be "." for a bare filename — ignore
    }
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS checkpoints (
        seq        INTEGER PRIMARY KEY AUTOINCREMENT,
        thread_id  TEXT    NOT NULL,
        run_id     TEXT    NOT NULL,
        state      TEXT    NOT NULL,
        ts         INTEGER NOT NULL
      );
      CREATE INDEX IF NOT EXISTS idx_checkpoints_thread
        ON checkpoints(thread_id, seq);
    `);
  }

  /** Append a state snapshot. Mutates `state.seq` and returns it. */
  save(state: AgentRunState): number {
    try {
      const info = this.db
        .prepare(
          "INSERT INTO checkpoints (thread_id, run_id, state, ts) VALUES (?, ?, ?, ?)",
        )
        .run(state.threadId, state.runId, JSON.stringify(state), Date.now());
      const seq = Number(info.lastInsertRowid);
      state.seq = seq;
      return seq;
    } catch (e) {
      log.warn({ err: e }, "checkpoint save failed — degrading to no-op");
      return -1;
    }
  }

  /** Latest checkpoint for a thread, or null. */
  loadLatest(threadId: string): AgentRunState | null {
    try {
      const row = this.db
        .prepare(
          "SELECT state FROM checkpoints WHERE thread_id = ? ORDER BY seq DESC LIMIT 1",
        )
        .get(threadId) as { state: string } | undefined;
      return row ? (JSON.parse(row.state) as AgentRunState) : null;
    } catch (e) {
      log.warn({ err: e }, "checkpoint loadLatest failed");
      return null;
    }
  }

  /** Exact snapshot at a sequence number (for rewind). */
  loadAt(seq: number): AgentRunState | null {
    try {
      const row = this.db
        .prepare("SELECT state FROM checkpoints WHERE seq = ?")
        .get(seq) as { state: string } | undefined;
      return row ? (JSON.parse(row.state) as AgentRunState) : null;
    } catch (e) {
      log.warn({ err: e }, "checkpoint loadAt failed");
      return null;
    }
  }

  /** All checkpoints for a thread, ascending (for replay/rewind menus). */
  loadThread(threadId: string): AgentRunState[] {
    try {
      const rows = this.db
        .prepare(
          "SELECT state FROM checkpoints WHERE thread_id = ? ORDER BY seq ASC",
        )
        .all(threadId) as { state: string }[];
      return rows.map((r) => JSON.parse(r.state) as AgentRunState);
    } catch (e) {
      log.warn({ err: e }, "checkpoint loadThread failed");
      return [];
    }
  }

  close(): void {
    try {
      this.db.close();
    } catch {
      /* ignore */
    }
  }
}
